bpo-11063: Fix _uuid module on macOS

vstinner opened the pull request:
On macOS, use uuid_generate_time() instead of
uuid_generate_time_safe() of libuuid, since uuid_generate_time_safe()
is not available.

<!-- issue-number: bpo-11063 -->
https://bugs.python.org/issue11063
<!-- /issue-number -->

vstinner committed:
bpo-11063: Fix _uuid module on macOS

On macOS, use uuid_generate_time() instead of
uuid_generate_time_safe() of libuuid, since uuid_generate_time_safe()
is not available.

vstinner commented:
I tested this PR manually on macOS Sierra: test_uuid pass, and is_safe attribute is unknown as expected.

```
macbook:master haypo$ ./python.exe -m test test_uuid 
Run tests sequentially
0:00:00 load avg: 0.75 [1/1] test_uuid
1 test OK.

Total duration: 337 ms
Tests result: SUCCESS

macbook:master haypo$ ./python.exe -c 'import uuid; print(uuid.uuid1().is_safe)'
SafeUUID.unknown
```

Note: I hacked _netstat_getnode() in Lib/uuid.py to add "-n" option to make tests faster, to prevent slow DNS resolution.

pitrou commented on Modules/_uuidmodule.c:
```diff
@@ -13,10 +19,23 @@ py_uuid_generate_time_safe(void)
+{
+    uuid_t out;
+    uuid_generate_time(out);
+    return PyBytes_FromStringAndSize((const char *) out, sizeof(out));
```
I'd prefer if you'd return the `(out, None)` tuple here.

vstinner replied:
I prefer to have very thin wrapper for C functions, and handle differences at the Python level.

pitrou replied:
IMO it only makes both the Python code and the C code more complicated.

pitrou replied:
One of the annoyances with modifying `uuid.py` is that there are so different code paths to take care of.

pitrou commented on Modules/_uuidmodule.c:
```diff
@@ -13,10 +19,23 @@ py_uuid_generate_time_safe(void)
+#ifdef HAVE_TIME_SAFE
     {"generate_time_safe", (PyCFunction) py_uuid_generate_time_safe, METH_NOARGS, NULL},
+#else
+    {"generate_time", (PyCFunction) py_uuid_generate_time, METH_NOARGS, NULL},
```
There's no need to expose two different function names, IMO.

vstinner replied:
> There's no need to expose two different function names, IMO.

It's not the same API:
* _uuid.generate_time_safe() returns (uuid, is_safe).
* _uuid.generate_time() only returns the uuid

pitrou replied:
I'm proposing that it's the same API (see above) :-)

vstinner commented:
I simplified my PR to not add a new function with a different name in _uuid. Instead, I modified uuid.py to decide if we have or not uuid_generate_time_safe() using sys.platform.

vstinner commented:
Instead of checking sys.platform, maybe we could have a constant in _uuid?

pitrou commented:
Yes, I think it's better to put this in `_uuid`.

vstinner commented:
> Yes, I think it's better to put this in _uuid.

Ok, done. I added _uuid.has_uuid_generate_time_safe (int: 1 or 0).

pitrou commented:
Thank you! :-)

warsaw commented:
I'd rather see a configure check for the existence of `uuid_generate_time_safe()` rather than hard coding it to platforms !APPLE for two reasons.  1) If macOS ever adds this API in some future release, this ifndef will be incorrect, and 2) if some other platform comes along that doesn't have this API, it will still use the incorrect function.

vstinner commented:
@warsaw: "I'd rather see a configure check for ..."

Please, go ahead and write it :-) I began to look at AC_CHECK_HEADER() but I don't see how to pass /usr/include/uuid/ to ask to test also this directory. Then I wanted to use pkg-config, but @pitrou showed me that it doesn't exist on macOS.

In short, modifying the autotools are out of my skills. I proposed a pragmatic solution to fix the _uuid module on macOS.

ned-deily commented:
I agree with @warsaw.  This should be a configure check.
